### ISSUE TITLE
OneTimeOnly refresh tokens are flagged as consumed before confirming it is received by the client.

### ISSUE THREAD

**Issue by LuizGsa21** [IdentityServer/IdentityServer4#5187]
OneTimeOnly refresh tokens are flagged as consumed before confirming it is received by the client. Mobile apps may drop requests at any moment. This racing condition becomes very apparent when having thousands of users.

Scenario:
1. App sees token is expired so it calls /connect/token providing a refresh_token.
2. Identity server receives refresh token request
3. Identity server flags refresh_token as consumed and generates new token + refresh_token.
4. App network fails (**or** app decides to give up on request do to slow network, **or** app is closed by user **or** an app crash occurs etc.... lol)
5. Now the app has a refresh_token that the Identity server assumes is consumed.

current code that flags token as consumed:
![image](https://user-images.githubusercontent.com/8405562/111018047-6ba39c80-837c-11eb-8f98-c64e5c01485b.png)

We should use the TCP layer to determine if response was received and then flag it as consumed. Something like:
![image](https://user-images.githubusercontent.com/8405562/111018135-da80f580-837c-11eb-9130-99917ae1b79f.png)

**Comment by leastprivilege** [comment]
From the docs 

> Adds a delegate to be invoked after the response has finished being sent to the client.

I don't think you can safely assume that just because the server is done sending, that the client actually received the response.

You can add a grace period to one-time only refreshing to allow a client to repeat the request within a certain time window. This is probably the more robust approach.

**Comment by LuizGsa21** [comment]
A write operation to a TCP socket will fail if the client fails to receive the data.

The sole purpose of the TCP layer is to make the connection reliable. The client reports to the server which packets it has received so the server knows which packets need to be resent in case some were lost on transit. Anyways, if the TCP layer says the client received the data, then it is safe to assume that it did.

After looking into ASP.NET core source code, it looks like they invoke the Response.OnComplete handler before writing to the socket… amazing.

basically:
1. invokes Response.OnComplete handler
2. writes pending data to socket
3. socket throws error so Response.OnComplete was falsely report.
![image](https://user-images.githubusercontent.com/8405562/111088506-8a796e80-84f5-11eb-877c-681448e787ea.png)

----
Also thanks for your suggestion. A 1 day grace period prob wouldn't hurt. Though there would be those special cases where If request fails and they dont open the app again within 24 hours they would have to login again.

**Comment by leastprivilege** [comment]
> A write operation to a TCP socket will fail if the client fails to receive the data.

Your are oversimplifying the internet. There are multiple hops between the client and server.

Either way - OneTime tokens are a security feature to protect against replay attacks - 24h seems to defeat the purpose. If you don't need/want the extra security, simply set the tokens to Re-use and don't bother with the complexity.

**Comment by LuizGsa21** [comment]
Yes I'm well aware that there are multiple hops on the internet. The TCP layer in the server's kernel speaks to the TCP layer in the clients kernel through those unreliable hops and make the connection reliable by speaking to each other on what was received/missed.

If you would like to learn more read TCP/IP Illustrated volume 1.

And thanks again for you suggestion. We decided to use reusable tokens and then use token invalidation endpoint.
